Add support for "show grants"

amrutagokhale opened the pull request:
This PR adds syntax support for "show grants" and its implementation for Hive connector. This command is not specified in SQL standard, so there there is no gold standard for the syntax. I have borrowed the syntax from Hive. Here is the syntax that the current version supports:
`SHOW GRANTS
        (USER identifier)?
        ON (TABLE? qualifiedName | ALL)`

It supports showing grants for users. I haven't tested the code for roles (I am planning to add support for roles later).
- When user name is speciified, `SHOW GRANTS USER identifier ...` will show grants for the specified user, provided the specified user is the same as the current user. Attempts to show grants for any other user than the current user will be denied.
- When user is not specified, `SHOW GRANTS ...` will show grants for the current user. 

The implementation follows the definition of `TABLE_PRIVILEGES` table specified in SQL standard:

``` sql
CREATE VIEW TABLE_PRIVILEGES AS
SELECT GRANTOR, GRANTEE, TABLE_CATALOG, TABLE_SCHEMA, TABLE_NAME,
PRIVILEGE_TYPE, IS_GRANTABLE, WITH_HIERARCHY
FROM DEFINITION_SCHEMA.TABLE_PRIVILEGES
WHERE ( GRANTEE IN
( ’PUBLIC’, CURRENT_USER )
OR
GRANTOR = CURRENT_USER )
AND
TABLE_CATALOG
= ( SELECT CATALOG_NAME
FROM INFORMATION_SCHEMA_CATALOG_NAME );
```

except that:
1) We don't consider the case of grantor = current_user (I couldn't get the information about grantor for the hive connector)
2) grantee in `PUBLIC` is not taken into consideration - I am currently testing whether the `PUBLIC` keyword works as intended for grant/revoke, and once that's being done, I will add code for `PUBLIC`.

Internal review at TD: https://github.com/Teradata/presto/pull/387 and https://github.com/Teradata/presto/pull/414

PS: I will add documentation soon once this gets in.

electrum commented on presto-spi/src/main/java/com/facebook/presto/spi/security/Privilege.java:
```diff
@@ -1,19 +0,0 @@
- */
-package com.facebook.presto.spi.security;
-
-public enum Privilege
```
Why not keep the enum as a top level class?

electrum replied:
It's a backwards incompatible change for seemingly no benefit. We avoid making breaking changes to the SPI when possible.

amrutagokhale replied:
Okay, I edited the commits to keep `Privilege` class as-is and added `PrivilegeInfo` as another class.

maciejgrzybek commented on presto-docs/src/main/sphinx/sql/show-grants.rst:
```diff
@@ -0,0 +1,42 @@
+
+Specifying ``ALL`` lists the grants on all the tables in the current catalog.
+
+The commands needs the current catalog to be set.
```
`need`

maciejgrzybek commented on presto-docs/src/main/sphinx/sql/show-grants.rst:
```diff
@@ -0,0 +1,42 @@
+
+List the grants for the current user on the specified table in the current catalog.
+
+Specifying ``ALL`` lists the grants on all the tables in the current catalog.
```
What happens if catalog is not set?
What about schema?
Is it possible to do a `show grants on hive.default.foo`?

amrutagokhale replied:
It is okay if schema is not set. If the catalog is not set however, the command gives an error. And yes, you can do `show grants on hive.default.foo` (assuming that catalog=hive has already been set).

maciejgrzybek replied:
What schema is used if none is given?
Does it list all tables in all schemas in given catalog?

amrutagokhale commented:
Here is the latest status on this PR:

* I changed the syntax so as to remove `USER username` part from the syntax ( having that as part of the syntax was of no use, since the only user for which anyone could see the grants was for the current user ). 
* In addition to showing the privileges granted directly to the user, the command will also show privileges that the user possesses as a result of the privileges granted to the roles that the user belongs. So  `grantee in PUBLIC` is taken into consideration (see my first comment to get the context).
* Added documentation.
* Rebased and resolved conflicts.

@electrum Can you take a look, please?

amrutagokhale commented:
Rebased and resolved conflicts.

cawallin commented on presto-main/src/main/java/com/facebook/presto/metadata/MetadataListing.java:
```diff
@@ -73,6 +74,18 @@ private MetadataListing() {}
+                .collect(Collectors.toSet());
+ */
+        Set<GrantInfo> grants = ImmutableSet.copyOf(metadata.listTablePrivileges(session, prefix, grantee));
+        return ImmutableSet.copyOf(accessControl.filterGrants(session.getRequiredTransactionId(), session.getIdentity(), prefix, grants));
```
No need for ImmutableSet.copyOf again, just the above one guarantees you're safe.

cawallin commented on presto-main/src/main/java/com/facebook/presto/security/AccessControl.java:
```diff
@@ -165,17 +167,21 @@
      */
     void checkCanRevokeTablePrivilege(TransactionId transactionId, Identity identity, Privilege privilege, QualifiedObjectName tableName);
 
+    void checkCanShowGrants(TransactionId transactionId, Identity identity, QualifiedTablePrefix prefix);
```
Add javadoc for these methods too

cawallin commented on presto-spi/src/main/java/com/facebook/presto/spi/security/AccessDeniedException.java:
```diff
@@ -228,14 +228,14 @@ public static void denyRevokeTablePrivilege(String privilege, String tableName,
     }
 
-    public static void denyShowGrants(String user, String extraInfo)
+    public static void denyShowGrants(String extraInfo)
```
How do you know that "on" would be the proper formatting for this?
Saying the user still seems helpful.

amrutagokhale replied:
At present, `SHOW GRANTS` doesn't have any provision for showing grants for users other than the current user. So the user is implicitly the current session user. If and when we add the ability to show grants for other users, we could user user information.

cawallin replied:
It looks like you're only passing in the table name as "extraInfo" -- can you change the variable name to "tableName"? It still doesn't make sense to add "on" for any arbitrary string (which is what extraInfo suggests)

cawallin commented on presto-main/src/main/java/com/facebook/presto/connector/informationSchema/InformationSchemaPageSourceProvider.java:
```diff
@@ -210,32 +211,27 @@ private InternalTable buildTables(Session session, String catalogName, Map<Strin
 
         InternalTable.Builder table = InternalTable.builder(informationSchemaTableColumns(TABLE_TABLE_PRIVILEGES));
         for (GrantInfo grant : grants) {
-            if (session.getIdentity().getUser().equals(grant.getIdentity().getUser())) {
```
It seems like you're losing this check?

cawallin replied:
You probably want to keep it just in this method, you don't want this check in as general a place as MetadataListing

amrutagokhale replied:
please see above https://github.com/prestodb/presto/pull/6629#discussion_r102848441

cawallin commented on presto-main/src/main/java/com/facebook/presto/metadata/MetadataListing.java:
```diff
@@ -73,6 +74,18 @@ private MetadataListing() {}
 
+    public static Set<GrantInfo> listTablePrivileges(Session session, Metadata metadata, AccessControl accessControl, QualifiedTablePrefix prefix, String grantee)
+    {
+/*        Set<PrivilegeInfo> grants = metadata.listTablePrivileges(session, prefix, grantee).stream()
```
You completely lose the 
 .filter(grantInfo -> grantInfo.getIdentity().getUser().equalsIgnoreCase(grantee))
Also, remove this comment.

amrutagokhale replied:
I don't lose it. That check is now redundant. I am extracting the grantee from the session via `extractGrantee(session, filters);` and passing that to `listTablePrivileges` method. So the privileges returned are only for the given session user.

cawallin commented on presto-main/src/main/java/com/facebook/presto/security/AccessControl.java:
```diff
@@ -178,8 +178,15 @@
     void checkCanRevokeTablePrivilege(TransactionId transactionId, Identity identity, Privilege privilege, QualifiedObjectName tableName);
 
+    /**
+     * check if identity is allowed to show grants on the specified catalog or schema or table.
```
nit: Capitalize "check" here and everywhere it's used

amrutagokhale commented:
@dain can you take a look at this? Would be great if this goes in first, since the PR about role management is built on top of this code.

amrutagokhale commented:
I added access control for `SHOW GRANTS` to the `AccessControl` interface, since this would be needed when adding system level access control such as in this one: https://github.com/prestodb/presto/pull/7695

dain commented on presto-spi/src/main/java/com/facebook/presto/spi/security/Privilege.java:
```diff
@@ -15,5 +15,5 @@
 public enum Privilege
 {
-    SELECT, DELETE, INSERT
+    SELECT, DELETE, INSERT, UPDATE
```
Is it possible to have an `UDATE` grant?  If not, I'd leave it out for now since we can't test it.

amrutagokhale replied:
If I don't add it here, `SHOW GRANTS` won't show the privilege, even if user possesses it, for ex., in the case of the owner of a Hive table who has all the privileges, including the `UPDATE` privilege. I can add a product test, testing whether `UPDATE` is shown for the owner of the table.

dain commented on presto-main/src/main/java/com/facebook/presto/sql/rewrite/ShowQueriesRewrite.java:
```diff
@@ -198,6 +200,38 @@ protected Node visitShowTables(ShowTables showTables, Void context)
+
+            Optional<QualifiedName> tableName = showGrant.getTableName();
+            if (tableName.isPresent()) {
+                QualifiedObjectName qualifiedTableName = createQualifiedObjectName(session, showGrant, tableName.get());
```
Do we need to validate that the table name is a valid table?

dain commented on presto-product-tests/src/main/java/com/facebook/presto/tests/hive/TestGrantRevoke.java:
```diff
@@ -101,7 +101,20 @@ public void testGrantRevoke()
+        assertThat(bobExecutor.executeQuery(format("SHOW GRANTS ON %s", tableName)))
+                .containsOnly(ImmutableList.of(
+                        row("bob", "hive", "default", "alice_owned_table", "SELECT", Boolean.TRUE),
+                        row("bob", "hive", "default", "alice_owned_table", "INSERT", Boolean.FALSE)
```
should `"alice_owned_table"` be `tableName`?

dain commented on presto-docs/src/main/sphinx/sql/show-grants.rst:
```diff
@@ -0,0 +1,46 @@
+
+List the grants for the current user on the specified table in the current catalog.
+
+Specifying ``ALL`` lists the grants for the current user on all the tables in all the schemas in the current catalog.
```
```md
Specifying ``ALL`` lists the grants for the current user on all the tables in all schemas of the current catalog.
```

martint commented on presto-parser/src/main/antlr4/com/facebook/presto/sql/parser/SqlBase.g4:
```diff
@@ -59,6 +59,8 @@ statement
         (GRANT OPTION FOR)?
         (privilege (',' privilege)* | ALL PRIVILEGES)
         ON TABLE? qualifiedName FROM grantee=identifier                #revoke
+    | SHOW GRANTS
```
Why `SHOW GRANTS ON ALL` for all grants instead of just `SHOW GRANTS`?

amrutagokhale replied:
I followed the same syntax that `SHOW GRANT` command in Hive has. But I guess, `ON ALL` seems redundant; I can remove it.

amrutagokhale commented on presto-parser/src/main/java/com/facebook/presto/sql/tree/ShowGrants.java:
```diff
@@ -0,0 +1,109 @@
+    private final Optional<QualifiedName> tableName;
+    private final boolean all;
+
+    public ShowGrants(boolean table, Optional<QualifiedName> tableName, boolean all)
```
Having the `TABLE` keyword is optional; you can specify `SHOW GRANTS ON TABLE <tablename>` or `SHOW GRANTS ON <tablename>`. So can't infer `table` flag. About the `all` flags, from what Martin suggested, I think we can get rid of it altogether.

cawallin replied:
Is the TABLE keyword necessary? Might be simpler to just remove it.

dain replied:
We we add support for grants on functions and other things, we will ned a way to differentiate them.

dain commented on presto-main/src/main/java/com/facebook/presto/connector/informationSchema/InformationSchemaPageSourceProvider.java:
```diff
@@ -202,6 +208,34 @@ private InternalTable buildTables(Session session, String catalogName, Map<Strin
+
+        InternalTable.Builder table = InternalTable.builder(informationSchemaTableColumns(TABLE_TABLE_PRIVILEGES));
+        for (GrantInfo grant : grants) {
+            if (session.getIdentity().getUser().equals(grant.getIdentity().getUser())) {
```
Not that we need to do this now, but eventually we should add `|| session.getIdentity().getUser().equals(grant.getGrantor().orElse(null))` here

amrutagokhale replied:
Good point. I will add it once we add information about the grantor (not exactly here though, since I changed this piece of code).

dain commented on presto-main/src/main/java/com/facebook/presto/connector/informationSchema/InformationSchemaPageSourceProvider.java:
```diff
@@ -202,6 +208,34 @@ private InternalTable buildTables(Session session, String catalogName, Map<Strin
+        return table.build();
+    }
+
+    private List<GrantInfo> getTablePrivilegesList(Session session, String catalogName, Map<String, NullableValue> filters)
```
I'd just inline this.

amrutagokhale replied:
Actually, I got rid of this in a later commit.

dain commented on presto-spi/src/main/java/com/facebook/presto/spi/security/AccessDeniedException.java:
```diff
@@ -228,6 +228,16 @@ public static void denyRevokeTablePrivilege(String privilege, String tableName,
         throw new AccessDeniedException(format("Cannot revoke privilege %s on table %s%s", privilege, tableName, formatExtraInfo(extraInfo)));
     }
 
+    public static void denyShowGrants(String user)
```
I think this is in the wrong commit

dain commented:
For this commit message:
```
Add access control for SHOW GRANTS to AccessControl interface

We don't need to add it on the connector level, since
SHOW GRANTS is only implemented for the hive connector
at present and it only shows grants for the current user, so
every user should be able to execute it.
```

I think it would be easier to understand if you replace "We don't" with a "Only add to system access control, since..`

dain commented on presto-main/src/main/java/com/facebook/presto/security/AccessControl.java:
```diff
@@ -165,18 +167,29 @@
+     * Check if identity is allowed to show grants on the specified catalog or schema or table.
+     * @throws com.facebook.presto.spi.security.AccessDeniedException if not allowed
+     */
+    void checkCanShowGrants(TransactionId transactionId, Identity identity, QualifiedTablePrefix prefix);
```
Since checking grants is really about reading Table metadata, I think we should reuse the existing `checkCanShowTables` and `filterTables` method.... I think it would be weird to be able to discover tables (e.g., with PUBLIC permissions) via the show grants command, that you can discover with show tables.

dain replied:
Also, when we added `SHOW TABLES` we wanted to only show tables you had grants on, which is how the spec says the information schema table it is based on should work, but didn't know how to do that at the time.  My point is these APIs are really linked (and it would be nice to align to the spec in the future).

amrutagokhale replied:
Ok, I could have reused `checkCanShowTables` method (perhaps not `filterTables` method, since what is being returned is different than in `filterGrants`). But then, we have to think about what error message to throw. `checkCanShowTables` throws "Cannot show tables..." message, which won't be right in this case. One option is to create a common method `checkCanShowTableMetadata` that will throw an error message along the lines of "Cannot show metadata for tables in ...". So considering the refactoring involved, I didn't change the access control methods for now. Thoughts?

dain replied:
We like the idea of renaming and changing the error message.

I still think we should reuse the filterTables method.

martint commented on presto-parser/src/main/antlr4/com/facebook/presto/sql/parser/SqlBase.g4:
```diff
@@ -60,7 +60,7 @@ statement
         (privilege (',' privilege)* | ALL PRIVILEGES)
         ON TABLE? qualifiedName FROM grantee=identifier                #revoke
     | SHOW GRANTS
-        ON (ALL | TABLE? qualifiedName)                                #showGrants
```
What's the expected behavior when the TABLE keyword is missing?

Is it the case that:
- "SHOW GRANTS ON TABLE xxx" only works if a table named xxx exists, but would fail for a view called xxx or some other entity we want to support in the future.
- "SHOW GRANTS ON xxx" would work for any entity named xxx (what if namespaces are not exclusive? e.g., functions vs tables)

dain replied:
IIRC tables and views are both managed with the `TABLE` keyword, but you question is valid if we consider grants on functions.

martint replied:
It would be weird for TABLE to apply both to tables and views given that we distinguish between them in other contexts (SHOW CREATE TABLE vs SHOW CREATE VIEW, DROP TABLE vs DROP VIEW)

dain replied:
That is the way the `GRANT` statement works in the spec https://ronsavage.github.io/SQL/sql-2003-2.bnf.html#grant statement

amrutagokhale replied:
@martint I am unclear on how namespaces will be resolved when there are two entities named xxx and someone issues "SHOW GRANTS ON xxx". Perhaps we can defer resolving this issue until we add support for grants on entities other than tables.

martint replied:
That's reasonable

amrutagokhale commented on presto-spi/src/main/java/com/facebook/presto/spi/security/GrantInfo.java:
```diff
@@ -0,0 +1,86 @@
+    private final Identity grantee;
+    private final SchemaTableName schemaTableName;
+    private final Optional<Identity> grantor;
+    private final Optional<Boolean> withHierarchy;
```
This denotes whether the privilege is automatically granted for all the derived objects too (e.g., on views derived from the table). Hive doesn't seem to keep this information, so I can't retrieve it. (also, I just noticed that there is an optional `WITH HIERARCHY OPTION` clause in `GRANT` statement as per the SQL standard, but Presto doesn't support it yet).

amrutagokhale commented on presto-main/src/main/java/com/facebook/presto/metadata/MetadataManager.java:
```diff
@@ -811,6 +813,27 @@ public void revokeTablePrivileges(Session session, QualifiedObjectName tableName
+                        .collect(toSet()));
+            }
+        }
+        return ImmutableList.copyOf(grants);
```
Changed that bit of code.

amrutagokhale commented:
@dain I have addressed your comments.

dain commented on presto-hive/src/main/java/com/facebook/presto/hive/HiveMetadata.java:
```diff
@@ -1239,6 +1241,28 @@ public void revokeTablePrivileges(ConnectorSession session, SchemaTableName sche
+            Set<PrivilegeInfo> privileges = metastore.getTablePrivileges(session.getUser(), tableName.getSchemaName(), tableName.getTableName()).stream()
+                    .map(HivePrivilegeInfo::toPrivilegeInfo)
+                    .flatMap(Set::stream)
+                    .distinct()
```
I don't think you need `distinct` with `toSet`.

dain commented on presto-hive/src/main/java/com/facebook/presto/hive/HiveMetadata.java:
```diff
@@ -1239,6 +1241,28 @@ public void revokeTablePrivileges(ConnectorSession session, SchemaTableName sche
+                    .map(HivePrivilegeInfo::toPrivilegeInfo)
+                    .flatMap(Set::stream)
+                    .distinct()
+                    .collect(toSet());
```
You can use `ImmutableSet.toImmutableSet()` now

dain commented:
I still think the commit message would be easier to read if it was not stated in the negative form "We don't need to add it on the connector level, since" and instead stated with something like ""Only add to system access control, since.."

dain commented:
We also need another look at the parser stuff from @martint

amrutagokhale commented:
@dain I have addressed your comments. I am now reusing the `checkCanShowTablesMetadata` for doing access control check for `SHOW GRANTS`. The order of commits shown on github isn't correct. `f3ada63abedd3cb084138a8b964d5d75b9706d66 (Rename checkCanShowTables to checkCanShowTablesMetadata)` should be far ahead of other commits, specifically before `0ef988789ff69fa9967f803ba0ec6f686bd91cd1 (Rewrite SHOW GRANTS as a SELECT query)`.

dain commented:
Merged, thanks!
